### PR TITLE
Closes #3569, #3689 Disable automatic paragraph filter for the Standard text format

### LINKED ISSUE #3569: Embedding any block in a text paragraph breaks linked images
### Describe the bug
A link may be added to an image embedded in a text paragraph by clicking on the image to select it and then clicking the **Link media** button in the pop-up media toolbar. See this issue regarding the implementation of this feature in Quickstart: [Allow displaying images as linked images #2366](https://github.com/az-digital/az_quickstart/issues/2366).

However, if a block is embedded in the same text area, the link for the image will not work on the published page. Any block causes this behavior, even an empty flexible block.

**Note:** For guidance about the accessible use of images, especially those that are linked, see https://www.w3.org/WAI/tutorials/images/.

### To Reproduce

1. When editing a page, add a new Text paragraph to a page by clicking the **Add Text** button.
2. Click the **Insert Media** button in the editor toolbar.
3. Select an image and insert it into the text area.
4. Click the image to select it.
5. Click the **Link media** button in the row of icons above/below the image.
6. Enter any link and click the green check mark button.
7. Click elsewhere in the text area to deselect the image (you may need to click the new line button on the image).
8. Click the **Embed Content Block** button in the editor toolbar.
9. Select any block, click the **Next** button, and then click the **Embed** button.
10. Save the page.
11. Observe that the image link does not work on the published page.

## Proposed resolution
This may be a CKEditor 5 issue. As a workaround, users can arrange their content to ensure that linked images are not embedded in the same text paragraph as a content block.

### Expected behavior

Linked images should be functional regardless of what other content is embedded in the same text paragraph.

## Additional context
When the link of the linked image works correctly, it appears that the HTML of that section is malformed: the `<p>` element containing the `<a>` element is closed before the `<a>` element is closed. Then, after the `<a>` element is closed, an extra unclosed `<p>` tag is added.

### PR DESCRIPTION
**Pull request by bberndt-uaz**
This PR proposes to address #3569 and #3689 by doing the following:

1. In the filters for the Standard text format (`az_standard`), disable the `Convert line breaks into HTML (i.e. <br> and <p>`) filter (with the machine name `filter_autop`).
   - Question: This PR does NOT disable the filter for the Restricted HTML or Plain text formats, since they do not use the CKEditor. Is this correct?
     - Yes. See @tadean's [response](#issuecomment-2430390898) below.
   - This PR does not remove the patch for [drupal.org issue 2350049](https://www.drupal.org/project/drupal/issues/2350049) since the `filter_autop()` function is used by our TextFormatRecognizer process plugin and it may be used elsewhere as well.
2. Reorder the filters in filter.format.az_standard.yml (a side effect of exporting the updated configuration).
3. In the Quickstart demo content, update the body text for News and Events to replace line breaks with the appropriate HTML.

#### Update Oct-22-2024:

Summary of the rationale for disabling the `filter_autop` text filter:
- The `filter_autop` filter is causing issues with the appearance and functionality of multiple elements in Quickstart. These issues can be resolved by simply disabling the filter; otherwise, the issues would require multiple targeted fixes, which could potentially break with future Drupal core updates.
- Disabling the filter is likely to have no effect on most pages of most Quickstart sites. Where unintended changes do occur, the impact is low risk: a small amount of vertical space will be added above or below an element.
- Disabling the filter is a solution [suggested by others in the Drupal community](https://www.drupal.org/project/drupal/issues/3075527#comment-15604529).
- Currently, with this filter enabled, the source code view of text area content may not reflect the final HTML of that text area. Disabling this filter will ensure editors that "what you see is what you get".
- Disabling the filter also fixes an issue where Quickstart pages can have invalid HTML (a closing `</p>` tag without an opening `<p>` tag).

#### Update Sep-18-2024:
Based on further testing of existing sites with this PR applied, disabling `filter_autop` in the Standard text format would have the following effects:

1. Fix the issue where linked images are broken when a block is embedded in the same text area
2. Fix extra spacing and incorrect styling in image credit/captions on embedded images
3. Remove extra `<p></p>` before and after certain embedded blocks
4. Remove extra `<p></p>` after certain embedded images
   - (This does not remove the 1 rem margin-bottom on `<figure>` elements from Bootstrap)
5. Preserve `<br><br>` instead of converting to `<p></p>`
6. Preserve an initial `<br>` within a `<p>` instead of removing it in some circumstances

#### Related Drupal core/contrib issues (as identified by @joeparsons):

 - [_filter_autop() breaks links around the tag](https://www.drupal.org/project/drupal/issues/3075527)
 - [Prevent drupal links being added to embedded entities that contain links when rendered](https://www.drupal.org/project/entity_embed/issues/3061449)

#### Code for the filter_autop function:

 - https://git.drupalcode.org/project/drupal/-/blob/10.3.x/core/modules/filter/filter.module?ref_type=heads#L694-787

### Release notes

> This release disables the `Convert line breaks into HTML (i.e. <br> and <p>`) filter for the Standard text format. This change resolves issues with the appearance and functionality of multiple elements in Quickstart. However, while most existing content should be unaffected, this change may result in some minor spacing changes, where extra blank lines may be removed or added. We may consider enabling this filter again in the future if [this Drupal core](https://www.drupal.org/project/drupal/issues/3075527) issue regarding the filter is resolved.

## Related issues

- #3569
- #3689

## How to test

Probo review site: https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-3696.probo.build

Any further testing on existing sites would be welcome! In a test or local environment, you can simply disable the `Convert line breaks into HTML (i.e. <br> and <p>`) filter in the Standard text format.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My change requires release notes.

### CODE REVIEW

**Review by trackleft — DISMISSED** [review]

**Review by joeparsons — APPROVED** [review]

**Comment by bberndt-uaz** [comment]
- Per discussion in AZ Digital, this would be a minor release change.
- @tadean is planning to do some further testing of this PR.
- As a follow-up, we may need to update az_migration to use HTML for line breaks instead of line break characters.

**Comment by tadean** [comment]
From the code of `_filter_autop`, it appears that the filter is intended to be run before the filter for embedded media (because `_filter_autop` contains exclusions for `drupal-media`). Currently it looks like we run it after embedded media. Changing the order does help the issue but does not completely resolve it. However, there's another problem:

`filter_autop` uses a list of elements and ignored elements to split its content into chunks and determine where to insert newlines and paragraph tags, and the list of block level elements is currently:
```
$block = '(?:table|thead|tfoot|caption|col|colgroup|tbody|tr|td|th|div|dl|dd|dt|ul|ol|li|pre|select|option|form|map|area|blockquote|address|math|input|p|h[1-6]|fieldset|legend|hr|article|aside|details|figcaption|figure|footer|header|hgroup|menu|nav|section|summary)';
```

I think the fact that `a` is not one of these tags, and is immediately followed by a tag that should be ignored (`drupal-media`) means that I think the chunk boundary causes a new paragraph to get generated by the filter. If the list of "block level elements" instead contains:

```
$block = '(?:a|table|thead|tfoot|caption|col|colgroup|tbody|tr|td|th|div|dl|dd|dt|ul|ol|li|pre|select|option|form|map|area|blockquote|address|math|input|p|h[1-6]|fieldset|legend|hr|article|aside|details|figcaption|figure|footer|header|hgroup|menu|nav|section|summary)';
```

Then it seems to understand that there should not be a new block level element between the link and the embed and fixes the issue. I'm not sure I'd necessarily propose that exact change as a fix, since `a` definitely isn't a block level element, but it does illustrate we might be able to patch this function to understand that there shouldn't be a wrapping block between anchor and embed chunks.

```
<a href="http://www.drupal.org">
  <div>
    <figure style="max-width: 360px;">
      <div class="field field--name-field-media-az-image field--type-image field--label-visually_hidden">
        <div class="field__label visually-hidden">Image</div>
        <div class="field__item">
          <img loading="lazy" class="img-fluid image-style-az-small" src="/sites/default/files/styles/az_small/public/placeduck10.jpg.webp?itok=qodjr7ns" width="360" height="360" alt="Anders And">
        </div>
      </div>
    </figure>
  </div>
</a>
```

**Comment by bberndt-uaz** [comment]
We'll plan to discuss next steps for this issue in an AZ Digital workshop.

**Comment by bberndt-uaz** [comment]
Next steps: we will look into a potential fix for `_filter_autop`, as described in the comment from @tadean above. This potential fix would be combined in Quickstart with switching the order of the "Convert line breaks into HTML" and "Embed media" filters in the Standard text format. We can still consider disabling the "Convert line breaks into HTML" filter if we would like to do so as a larger improvement to Quickstart.

**Comment by bberndt-uaz** [comment]
I created an alternate pull request which partially resolves #3689:

 - #3806 

The alternate PR could be used as a temporary fix; however, as a long-term solution I'm still in favor of disabling filter_autop for the Standard text format. I added some notes on the rationale for this in the pull request description above.

**Comment by tadean** [comment]
>    * Question: This PR does NOT disable the filter for the Restricted HTML or Plain text formats, since they do not use the CKEditor. Is this correct?

It seems so to me, I think.. With these particular text formats, that filter is the only easy way to create a new line, since whitespace they enter in these text formats wouldn't otherwise create any line spacing for the viewer.

File: modules/custom/az_demo/data/az_demo_event_node.json [hunk 1]
```diff
@@ -213,8 +213,8 @@
     "az_event_date" : "today + 1 days 18:00",
     "az_event_date_to" : "today + 1 days 20:00",
     "az_event_date_duration" : "custom",
-    "az_summary" : "Screening: The Adventures of Picasso. https://example.com/The_Adventures_of_Picasso",
-    "az_body" : "Directed by Tage Danielsson, 1978 (original title: Picassos äventyr).\n\nIn a fantastically re-imagined biography of Pablo Picasso, the famous painter encounters many of the artistic figures and political events of the twentieth century.",
+    "az_summary" : "Screening: The Adventures of Picasso.",
```
Comment (bberndt-uaz) [inline modules/custom/az_demo/data/az_demo_event_node.json:216]:
```suggestion
    "az_summary" : "Screening: The Adventures of Picasso. https://example.com/The_Adventures_of_Picasso",
```

**Review by trackleft — APPROVED** [review]

**Review by tadean — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]

**Pull request merged**